diffprivlib/models/forest.py

```diff
@@ -0,0 +1,598 @@
+        self.fitted_ = True
+
+        return self
+
+    def predict(self, X):
+        check_is_fitted(self)
+
+        y = []
+        X = np.array(X)
+        check_array(X)
+
+        for x in X:
+            node = self.tree_.classify(x)
+            y.append(node.noisy_label)
+
+        return np.array(y)
+
+
+class DecisionNode:
+    """Base Decision Node
+    """
+    CONT_SPLIT = 0
+    CAT_SPLIT = 1
+
+    def __init__(self, level, split_feature=None, split_value=None, split_type=None):
+        self._level = level
+        self._split_type = split_type
+        self._split_feature = split_feature
+        self._split_value = split_value
+        self._left_child = None
+        self._right_child = None
+        self._cat_children = {}
+        self._class_counts = defaultdict(int)
+        self._noisy_label = None
+        self._sensitivity = -1.0
+
+    @property
+    def noisy_label(self):
+        return self._noisy_label
+
+    def set_split_value(self, split_value):
+        self._split_value = split_value
+
+    def set_split_type(self, split_type):
+        self._split_type = split_type
+
+    def set_left_child(self, node):
+        self._left_child = node
+
+    def set_right_child(self, node):
+        self._right_child = node
+
+    def add_cat_child(self, cat_value, node):
+        self._cat_children[str(cat_value)] = node
+
+    def is_leaf(self):
+        return not self._left_child and not self._right_child and not self._cat_children
+
+    def update_class_count(self, class_value):
+        self._class_counts[class_value] += 1
+
+    def classify(self, x):
+        if self.is_leaf():
+            return self
+
+        child = None
+
+        if self._split_type == self.CAT_SPLIT:
+            x_val = str(x[self._split_feature])
+            child = self._cat_children.get(x_val)
+        else:
+            x_val = x[self._split_feature]
+            if x_val < self._split_value:
+                child = self._left_child
+            else:
+                child = self._right_child
+
+        if child is None:
+            return self
+
+        return child.classify(x)
+
+    def set_noisy_label(self, epsilon, class_values):
+        if self.is_leaf():
+            if not self._noisy_label:
+                for val in class_values:
+                    if val not in self._class_counts:
+                        self._class_counts[val] = 0
+
+                if max([v for k, v in self._class_counts.items()]) < 1:
+                    self._noisy_label = np.random.choice([k for k, v in self._class_counts.items()])
+                else:
+                    all_counts = sorted([v for k, v in self._class_counts.items()], reverse=True)
+                    count_difference = all_counts[0] - all_counts[1]
+                    self._sensitivity = np.exp(-1 * count_difference * epsilon)
+                    self._noisy_label = self._calc_label_by_expo_mech(epsilon, self._sensitivity, self._class_counts)
```

Comment:
To implement Sam's changes, you can change these lines to the following, and delete the `_calc_label_by_expo_mech` method.
```
utility = list(self._class_counts.values())
candidates = list(self._class_counts.keys())
mech = PermuteAndFlip(epsilon=epsilon, sensitivity=1, monotonic=True, utility=utility, candidates=candidates)
self.noisy_label_ = mech.randomise()
```

Reply:
Thanks, Naoise. I will add that in.

Reply:
@naoise-h is there any risk of exponential floating point errors when class counts are high? Is it worth subtracting the smallest class count from each count, so that we baseline the utilities at zero? 

Not sure if its a problem in practice though.

Reply:
You're correct Sam, but this is handled within the mechanism itself, here:

https://github.com/IBM/differential-privacy-library/blob/main/diffprivlib/mechanisms/exponential.py#L235